src/video/wayland/SDL_waylandwindow.c

```diff
@@ -448,8 +491,102 @@ static const struct xdg_toplevel_listener toplevel_listener_xdg = {
     handle_close_xdg_toplevel
 };
 
+#ifdef HAVE_LIBDECOR_H
+static void
+decoration_frame_configure(struct libdecor_frame *frame,
+                           struct libdecor_configuration *configuration,
+                           void *user_data)
+{
+    SDL_WindowData *wind = user_data;
+    SDL_Window *window = wind->sdlwindow;
+    int width, height;
+    enum libdecor_window_state window_state;
+    struct libdecor_state *state;
+
+    /* window size */
+    if (!libdecor_configuration_get_content_size(configuration, frame, &width, &height)) {
+        width = window->w;
+        height = window->h;
+    }
 
+    wind->resize.width = width;
+    wind->resize.height = height;
+
+    wind->resize.pending = SDL_TRUE;
+    wind->resize.configure = SDL_TRUE;
+    Wayland_HandlePendingResize(window);
+    wind->shell_surface.libdecor.initial_configure_seen = SDL_TRUE;
+
+    window->w = wind->resize.width;
+    window->h = wind->resize.height;
+
+    /* window state */
+    if (!libdecor_configuration_get_window_state(configuration, &window_state)) {
+        window_state = LIBDECOR_WINDOW_STATE_NONE;
+    }
+
+   /* Always send maximized/restored/focus events; if the event is redundant it will
+     * automatically be discarded (see src/events/SDL_windowevents.c).
+     *
+     * No, we do not get minimize events from libdecor.
+     */
+    if (!(window_state & LIBDECOR_WINDOW_STATE_FULLSCREEN)) {
+        window->flags &= ~SDL_WINDOW_FULLSCREEN_DESKTOP;
+        SDL_SendWindowEvent(window,
+                            (window_state & LIBDECOR_WINDOW_STATE_MAXIMIZED) ?
+                                SDL_WINDOWEVENT_MAXIMIZED :
+                                SDL_WINDOWEVENT_RESTORED,
+                            0, 0);
+    } else {
+        window->flags |= SDL_WINDOW_FULLSCREEN_DESKTOP;
+    }
```

Comment:
I was able to confirm that @cgutman's fullscreen fix is still required for xdg-shell and is also required for libdecor, both on GNOME and Plasma. This patch takes care of everything, tested with SOR4 using a wedged-in key to test SDL_MinimizeWindow:

```suggestion
    if (window_state & LIBDECOR_WINDOW_STATE_FULLSCREEN) {
        window->flags |= SDL_WINDOW_FULLSCREEN_DESKTOP;
    } else {
        if (window->flags & SDL_WINDOW_FULLSCREEN_DESKTOP) {
            /* We might need to re-enter fullscreen after being restored from minimized */
            SDL_WaylandOutputData *driverdata = (SDL_WaylandOutputData *) SDL_GetDisplayForWindow(window)->driverdata;
            SetFullscreen(window, driverdata->output);
        } else {
            SDL_SendWindowEvent(window,
                                (window_state & LIBDECOR_WINDOW_STATE_MAXIMIZED) ?
                                    SDL_WINDOWEVENT_MAXIMIZED :
                                    SDL_WINDOWEVENT_RESTORED,
                                0, 0);
        }
    }
```

Reply:
It seems that after this, I am getting fullscreen issues with Capsized again. The SDL demos and Capsized used to work with the state before this, including transitioning between fullscreen and floating mode.

@flibitijibibo Can you confirm that you see the Capsized issues again?

Reply:
It looks like there's a bug somewhere in libdecor - it's claiming fullscreen state when it's not, possibly vice versa. The easiest test is to set the game to fullscreen and simply alt-tab, it will shrink to windowed mode even though it should still be in fullscreen mode (just in the background instead of in front). This appears to be unique to libdecor, and the configure callback we get on alt-tab claims that fullscreen is the current state when it is not. Interestingly, even if I shove a SetFullscreen call into the `LIBDECOR_WINDOW_STATE_FULLSCREEN` path it still doesn't work.

Aside from that it seems exactly as broken and inconsistent as it was before on my PC, except for resizable windows where this is now working correctly (SOR4 is the go-to example).

Reply:
Well, going straight into fullscreen when starting Capsized always worked. But it does not work after this change.

If I undo this change, the fullscreen mode works again with Capsized. Reading the raw wayland log (WAYLAND_DEBUG=1) I see that `set_fullscreen` is called twice now: Once right after initialisation, and the second time right after the configure event that follows the fullscreen event. That seems wrong.

I think the configure event is a result of requesting fullscreen. The configure event is only to inform the client that the surface properties changed. E.g. it carries the fullscreen flag because the user or compositor requested the fullscreen mode and the server is setting fullscreen mode and informing the client. It is not asking the client to request fullscreen mode again. In the configure callback, we should only update the internal state flags, send out notifications and redraw new content if the size changed. But this should not request the same mode (maximise, fullscreen, ...) again.

Reply:
It sounds like everything is really out of sync then - I would double check everything with xdg-shell first just to see what the existing behavior is like. At least on GNOME 40 fullscreen seems okay (minus the alt tab behavior).

Reply:
An initial test with xdg-shell shows that the min/max limits are not recognised so that the window will be resizeable. You can check this by pressing Alt+Space, e.g.with the `testscale` example, and see that the "Maximise" option is available. With libdecor this is greyed out (as it should) because the window is defined as non-maximisable. Taking these limits into account make it harder to switch between modes (floating, maximised, fullscreen).

In any case, I don't think there is a requirement to mimic the xdg-shell code paths. Both ways should be close to the X11 behaviour, minus all the things that are intentionally not possible with Wayland (e.g. setting the window position). That may involve different code paths.

Reply:
> In any case, I don't think there is a requirement to mimic the xdg-shell code paths. Both ways should be close to the X11 behaviour, minus all the things that are intentionally not possible with Wayland (e.g. setting the window position). That may involve different code paths.

This is true. I suppose at this point I would add SDL_MinimizeWindow during fullscreen to the list of tests, @cgutman may have one but I just shoved it into a local FNA build. Alt-tabbing back should return it to fullscreen mode.

Reply:
The patch removed the unsetting of the fullscreen flag via ` window->flags &= ~SDL_WINDOW_FULLSCREEN_DESKTOP;`. This caused the wrong fullscreen mode.